src/SAWScript/Crucible/LLVM/X86.hs

```diff
@@ -0,0 +1,514 @@
+  where
+    argRegs :: [Register]
+    argRegs = [Macaw.RDI, Macaw.RSI, Macaw.RDX, Macaw.RCX, Macaw.R8, Macaw.R9]
+    setRegSetupValue rs (reg, sval) = do
+      let ?ptrWidth = knownNat @64
+      val <- C.LLVM.unpackMemValue sym (C.LLVM.LLVMPointerRepr $ knownNat @64)
+        =<< resolveSetupVal cc mem env tyenv nameEnv sval
+      setReg reg val rs
+
+--------------------------------------------------------------------------------
+-- ** Postcondition
+
+-- | Assert the postcondition for the spec, given the final memory and register map.
+assertPost ::
+  Sym ->
+  LLVMCrucibleContext LLVMArch ->
+  Map MS.AllocIndex Ptr ->
+  MS.CrucibleMethodSpecIR LLVM ->
+  (Mem, Mem) {- ^ The state of memory before and after simulation -} ->
+  (Regs, Regs) {- ^ The state of the registers before and after simulation -} ->
+  IO ()
+assertPost sym cc env ms (premem, postmem) (preregs, postregs) = do
+  let ?ptrWidth = knownNat @64
+  let tyenv = ms ^. MS.csPreState . MS.csAllocs
+      nameEnv = MS.csTypeNames ms
+  forM_ (ms ^. MS.csPostState . MS.csPointsTos)
+    $ assertPointsTo sym cc env tyenv nameEnv postmem
+
+  prersp <- getReg Macaw.RSP preregs
+  expectedIP <- C.LLVM.doLoad sym premem prersp (C.LLVM.bitvectorType 8)
+    (C.LLVM.LLVMPointerRepr $ knownNat @64) C.LLVM.noAlignment
+  actualIP <- getReg Macaw.X86_IP postregs
+  correctRetAddr <- C.LLVM.ptrEq sym C.LLVM.PtrWidth actualIP expectedIP
+  C.addAssertion sym . C.LabeledPred correctRetAddr . C.SimError W4.initializationLoc
+    $ C.AssertFailureSimError "Instruction pointer not set to return address" ""
+
+  stack <- C.LLVM.doPtrAddOffset sym premem prersp =<< W4.bvLit sym knownNat 8
+  postrsp <- getReg Macaw.RSP postregs
+  correctStack <- C.LLVM.ptrEq sym C.LLVM.PtrWidth stack postrsp
+  C.addAssertion sym . C.LabeledPred correctStack . C.SimError W4.initializationLoc
+    $ C.AssertFailureSimError "Stack not preserved" ""
+
+  -- TODO: Match return value in RAX
+
+-- | Assert that a points-to postcondition holds.
+assertPointsTo ::
+  Sym ->
+  LLVMCrucibleContext LLVMArch ->
+  Map MS.AllocIndex Ptr {- ^ Associates each AllocIndex with the corresponding allocation -} ->
+  Map MS.AllocIndex LLVMAllocSpec {- ^ Associates each AllocIndex with its specification -} ->
+  Map MS.AllocIndex C.LLVM.Ident {- ^ Associates each AllocIndex with its name -} ->
+  Mem ->
+  LLVMPointsTo LLVMArch {- ^ crucible_points_to statement from the precondition -} ->
+  IO ()
+assertPointsTo sym cc env tyenv nameEnv mem (LLVMPointsTo _ tptr texpected) = do
+  let ?ptrWidth = knownNat @64
+  ptr <- C.LLVM.unpackMemValue sym (C.LLVM.LLVMPointerRepr $ knownNat @64)
+    =<< resolveSetupVal cc mem env tyenv Map.empty tptr
+  storTy <- C.LLVM.toStorableType =<< typeOfSetupValue cc tyenv nameEnv texpected
+  _actual <- C.LLVM.assertSafe sym =<< C.LLVM.loadRaw sym mem ptr storTy C.LLVM.noAlignment
+  -- TODO: actually do matching between actual and expected
```

Comment:
is it possible to check that the name of the expected starts with `_`?

Reply:
I added a warning for matching against things that aren't fresh variables starting with underscores.